### PR TITLE
Provide additional constraints in `shelleyBasedEraConstraints`

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    Provide additional constraints in `shelleyBasedEraConstraints`.  This will obsolete the following:
    - `obtainEraCryptoConstraints`
    - `obtainCryptoConstraints`
    - `obtainEraPParamsConstraint`
    - `obtainSafeToHashConstraint`
    Also provide additional constraints in `conwayEraOnwardsConstraints` and `shelleyToBabbageEraConstraints`
  compatibility: compatible
  type: feature
```

# Context

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-api/internal/Cardano/Api/Eras.hs [hunk 1]
```diff
@@ -623,7 +624,16 @@ cardanoEraConstraints = \case
 
 shelleyBasedEraConstraints :: ()
   => ShelleyBasedEra era
-  -> (Typeable era => IsShelleyBasedEra era => a)
+  -> (()
+      => Typeable era
```
Comment (carbolymer) [inline cardano-api/internal/Cardano/Api/Eras.hs:628]: With `ShelleyLedgerEra era ~ ledgerera` it would make constraints a little bit more readable.

**Review by carbolymer — APPROVED** [review]

File: cardano-api/internal/Cardano/Api/Eras.hs (hunk 1 above)
Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Eras.hs]: Done!

File: cardano-api/internal/Cardano/Api/Certificate.hs [hunk 2]
```diff
@@ -831,5 +827,5 @@ conwayCertificateConstraints
       , Ledger.TxCert (ShelleyLedgerEra era) ~ Ledger.ConwayTxCert (ShelleyLedgerEra era)
       ) => a)
   -> a
-conwayCertificateConstraints ConwayEraOnwardsConway f = f
+conwayCertificateConstraints = conwayEraOnwardsConstraints
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/Certificate.hs:830]: We can use a deprecation pragma. Not vital to do in this PR but when we want to remove it, let's use a deprecation pragma first.

File: cardano-api/internal/Cardano/Api/Feature/ShelleyBasedEra.hs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module Cardano.Api.Features.ShelleyBasedEra
```
Comment (Jimbo4350) [inline cardano-api/internal/Cardano/Api/Feature/ShelleyBasedEra.hs:1]: Accident?

Comment (newhoggy) [reply cardano-api/internal/Cardano/Api/Feature/ShelleyBasedEra.hs:1]: Yep.  Thanks!

**Pull request merged**